src/services/Payments.php

```diff
@@ -269,40 +312,144 @@ public function getBasketPayload(Order $order)
+				'order' => $order,
+				'adjustment' => $adjustment,
+				'amount' => Currency::formatAsCurrency($adjustment->amount, $order->paymentCurrency, $gateway->convertAmount, false, true),
 
-			//Convert to payment currency
-			$converted = Currency::formatAsCurrency($amount, $order->paymentCurrency, $gateway->convertAmount, false, true);
+			]);
+			$this->trigger(self::EVENT_BEFORE_BASKET_ADJUSTMENT_AMOUNT, $event);
+			$amount = $event->amount;
 
 			$lines[] = [
 				'qty' => 1,
 				'item_no' => $adjustment->id,
 				'item_name' => $adjustment->name,
-				'item_price' => $converted,
+				'item_price' => $amount * 100,
 				'vat_rate' => $taxrate
 			];
 		}
 
 		return $lines;
 	}
 
-	public function getShippingPayload(Order $order)
+	/**
+	 * Get the basket payload required for klarna to work
+	 *
+	 * @param Order $order
+	 * @return array
+	 */
+	private function _getKlarnaBasketPayload(Order $order): array
+	{
+		$lines = [];
+		$adjustments = [];
+		$taxrate = Plugin::getInstance()->getTaxes()->getProductTaxRate($order);
+		$subtract = 0;
+		$gateway = $order->getGateway();
+
+
+		//* Adjustments
+		foreach ($order->adjustments as $adjustment) {
+			// If adjustment is included, or shipping, skip
+			//? Included adjustments are already included in the line items
+			//? Shipping is added in the shipping payload
+			if ($adjustment->included || $adjustment->type == 'shipping') {
+				continue;
+			}
+
+			$event = new BasketAdjustmentAmount([
+				'order' => $order,
+				'adjustment' => $adjustment,
+				'amount' => Currency::formatAsCurrency($adjustment->amount, $order->paymentCurrency, $gateway->convertAmount, false, true),
+
+			]);
+			$this->trigger(self::EVENT_BEFORE_BASKET_ADJUSTMENT_AMOUNT, $event);
+			$amount = $event->amount;
+
+			// If adjustment is negative, add up
+			//? Klarna cannot handle negative values, therefore we need to subtract all negative values from the lineitems
+			if ($adjustment->amount < 0) {
+				$subtract = abs($subtract) + abs($amount);
+				continue;
+			}
+
+			// Add to adjustments array
+			$adjustments[] = [
+				'qty' => 1,
+				'item_no' => $adjustment->id,
+				'item_name' => $adjustment->name,
+				'item_price' => $amount * 100,
+				'vat_rate' => $taxrate
+			];
+		}
+
+		//* Lineitems
+		foreach ($order->getLineItems() as $lineItem) {
+
+			$event = new BasketLineTotal([
+				'order' => $order,
+				'lineItem' => $lineItem,
+				'total' => Currency::formatAsCurrency($lineItem->total, $order->paymentCurrency, $gateway->convertAmount, false, true),
+
+			]);
+			$this->trigger(self::EVENT_BEFORE_BASKET_LINE_TOTAL, $event);
+			$amount = $event->total;
+
+			// Subtract negative values from lineitems
+			//? We will subtract as much as posible from each item, not going below 0
+			if ($subtract > 0) {
+				if ($amount >= $subtract) {
+					$amount = $amount - $subtract;
+					$subtract = 0;
+				} else {
+					$subtract -= $amount;
+					$amount = 0;
+				}
+			}
```

Comment:
Avoid nesting ifs. A typical approach would be to refactor into a separate function. 

Example here: 
https://onlinephp.io/c/ebcbf